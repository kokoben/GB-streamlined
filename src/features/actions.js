export const FEATURE_VIDEO_SET = 'FEATURE_VIDEO_SET';
export const FEATURE_VIDEO_SET_SUCCESS = 'FEATURE_VIDEO_SET_SUCCEEDED';
export const FEATURE_VIDEO_SET_FAIL = 'FEATURE_VIDEO_SET_FAILED';
export const FEATURE_VIDEOS_SET = 'FEATURE_VIDEOS_SET';
export const FEATURE_VIDEOS_SET_SUCCESS = 'FEATURE_VIDEOS_SET_SUCCEEDED';
export const FEATURE_VIDEOS_SET_FAIL = 'FEATURE_VIDEOS_SET_FAILED';
export const LATEST_FEATURE_VIDEO_SET = 'LATEST_FEATURE_VIDEO_SET';
export const LATEST_FEATURE_VIDEO_SET_SUCCESS = 'LATEST_FEATURE_VIDEO_SET_SUCCEEDED';
export const LATEST_FEATURE_VIDEO_SET_FAIL = 'LATEST_FEATURE_VIDEO_SET_FAILED';

export const setFeatureVideo = (id) => {
  console.log('inside setFeatureVideo');
  return {
    type: FEATURE_VIDEO_SET,
    id
  }
}

export const setFeatureVideos = (page, pageSize) => {
  console.log('inside setFeatureVideos');
  return {
    type: FEATURE_VIDEOS_SET,
    page,
    pageSize
  }
}

export const setLatestFeatureVideo = () => ({
  type: LATEST_FEATURE_VIDEO_SET
})
