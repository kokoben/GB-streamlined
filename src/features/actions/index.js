import * as fActions from './types';

export const setFeatureVideo = (id) => {
  console.log('inside setFeatureVideo');
  return {
    type: fActions.FEATURE_VIDEO_SET,
    id
  }
}

export const setFeatureVideos = (page, pageSize) => {
  console.log('inside setFeatureVideos');
  return {
    type: fActions.FEATURE_VIDEOS_SET,
    page,
    pageSize
  }
}

export const setLatestFeatureVideo = () => ({
  type: fActions.LATEST_FEATURE_VIDEO_SET
})
