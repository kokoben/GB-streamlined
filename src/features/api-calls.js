import apiKey from '../api-keys';

// return list of all feature videos by publish date in descending order
/* eslint-disable quotes, import/prefer-default-export */
export const requestFeatureVideos = page => ([
  `https://www.giantbomb.com/api/videos/?api_key=${apiKey}`,
  `&format=jsonp&json_callback=callback&limit=24&offset=${(page - 1) * 24}`,
  `&field_list=hd_url,name,id,deck,embed_player,publish_date,image,video_type,user`,
  `&filter=video_type:8&sort=publish_date:desc`,
].join(''));
/* eslint-enable */
