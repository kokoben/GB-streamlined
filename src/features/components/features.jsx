import React from 'react';
import CurrentFeatureVideo from '../containers/current-feature-video';
import FeatureVideos from '../containers/feature-videos';
import FeatureFeedHeader from '../components/feature-feedheader';

export default () => (
  <div>
    <CurrentFeatureVideo />
    <FeatureFeedHeader />
    <FeatureVideos />
  </div>
);
