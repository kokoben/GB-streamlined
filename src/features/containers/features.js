import React from 'react';
import CurrentFeatureVideo from './current-feature-video';
import FeatureVideos from './feature-videos';

export default () => (
  <div>
    <CurrentFeatureVideo />
    <FeatureVideos />
  </div> 
)
