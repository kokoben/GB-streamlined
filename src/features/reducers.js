import { 
  FEATURE_VIDEO_SET_SUCCESS,
  FEATURE_VIDEO_SET_FAIL 
} from './actions';

export const featureVideo = (state=null, action) => {
  switch(action.type) {
    case FEATURE_VIDEO_SET_SUCCESS:
      console.log('feature_video_set_success!');
      const videoLink = action.response.results[0];
      return videoLink;
    case FEATURE_VIDEO_SET_FAIL:
        console.log('feature_video_set_fail!');
        console.log(action.message);
    default:
      return state;
  }
}
