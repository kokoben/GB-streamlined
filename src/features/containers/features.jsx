import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import CurrentFeatureVideoConnected from '../containers/current-feature-video';
import FeatureVideosConnected from '../containers/feature-videos';
import FeatureFeedHeaderConnected from '../components/feature-feedheader';
import { setSearchPage, setSearchMarker } from '../../actions/index';

export class Features extends Component {
  constructor(props) {
    super(props);

    this.onClickPageScroll = this.onClickPageScroll.bind(this);
  }

  componentDidMount() {
    this.props.setSearchPage(null);
    this.props.setSearchMarker(null);
  }

  onClickPageScroll() {
    this.feedHeaderNode.scrollIntoView();
  }

  render() {
    return (
      <div>
        <CurrentFeatureVideoConnected />
        <div ref={node => this.feedHeaderNode = node}>
          <FeatureFeedHeaderConnected />
        </div>
        <FeatureVideosConnected onClickPageScroll={this.onClickPageScroll} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => (
  bindActionCreators({
    setSearchPage,
    setSearchMarker,
  }, dispatch)
);

Features.propTypes = {
  setSearchPage: PropTypes.func.isRequired,
  setSearchMarker: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(Features);

