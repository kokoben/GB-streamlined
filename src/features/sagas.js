import { takeEvery, call, put } from 'redux-saga/effects';
import jsonp from 'jsonp-promise';
import * as fActions from './actions/types';
import { requestFeatureVideos } from './api-calls';
import { requestVideo } from '../api-calls';

const jsonpWrapper = args => {
  const response = jsonp(...args);
  return response.promise;
}

const params = {param: 'json_callback'};

// workers
function* setFeatureVideoAsync(action) {
  try {
    const jsonpArgs = [requestVideo(action.id), params];
    const response = yield call(jsonpWrapper, jsonpArgs);
    yield put({type: fActions.FEATURE_VIDEO_SET_SUCCESS, response: response.results});
  } catch (e) {
    console.log('setFeatureVideoAsync request failed!');
    yield put({type: fActions.FEATURE_VIDEO_SET_FAIL, message: e.message})
  }
}

function* setLatestFeatureVideoAsync() {
  try {
    const jsonpArgs = [requestFeatureVideos(1), params];
    const response = yield call(jsonpWrapper, jsonpArgs);
    yield put({type: fActions.FEATURE_VIDEO_SET_SUCCESS, response: response.results[0]});
  } catch (e) {
    yield put({type: fActions.FEATURE_VIDEO_SET_FAIL, message: e.message});
  }  
}

function* setFeatureVideosAsync(action) {
  try {
    const jsonpArgs = [requestFeatureVideos(action.page), params];
    const response = yield call(jsonpWrapper, jsonpArgs);
    yield put({type: fActions.FEATURE_VIDEOS_SET_SUCCESS, response: response})
  } catch (e) {
    console.log('feature_videos_set_failed!');
    yield put({type: fActions.FEATURE_VIDEOS_SET_FAIL, message: e.message});
  }
}

// watchers
export function* watchSetFeatureVideo() {
  console.log('redux-saga is running the FEATURE_VIDEO_SET action listener');
  yield takeEvery(fActions.FEATURE_VIDEO_SET, setFeatureVideoAsync);
}

export function* watchSetLatestFeatureVideo() {
  console.log('redux-saga is running the LATEST_FEATURE_VIDEO_SET action listener');
  yield takeEvery(fActions.LATEST_FEATURE_VIDEO_SET, setLatestFeatureVideoAsync); 
}

export function* watchSetFeatureVideos() {
  console.log('redux-saga is running the FEATURE_VIDEOS_SET action listener');
  yield takeEvery(fActions.FEATURE_VIDEOS_SET, setFeatureVideosAsync);
}
