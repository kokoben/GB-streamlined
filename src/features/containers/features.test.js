import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Features } from './features';

it('renders correctly', () => {
  const mockSetSearchPage = jest.fn();
  const wrapper = shallow(<Features
    setSearchPage={mockSetSearchPage}
  />);

  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});
