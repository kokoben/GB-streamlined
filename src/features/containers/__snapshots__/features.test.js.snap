// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <Connect(CurrentFeatureVideo) />
  <Connect(Component) />
  <Connect(FeatureVideos) />
</div>
`;
